### ISSUE TITLE
Onsite - Possible race condition

### ISSUE THREAD

**Issue by k-macmillan** [department-of-veterans-affairs/notification-api#718]
## Value Statement

The Onsite call was nested into a Celery Chain, which calls one after another. The celery chain is called with [apply_async()](https://flask-api.readthedocs.io/en/latest/_autosummary/app.celery.ContextTask.apply_async.html). Onsite is dependent on the return value from the `lookup_va_profile_id` method. This may be [the simplest method](https://docs.celeryq.dev/en/v4.4.7/userguide/calling.html#linking-callbacks-errbacks), but please review and determine the best route to move forward.

## Checklist

- [ ] The Onsite method gets the va profile id (if there's one to get)
- [ ] Modifications to tests if needed

## Additional Info/Resources

- https://docs.celeryq.dev/en/v4.4.7/userguide/index.html
- https://docs.celeryq.dev/en/v4.4.7/userguide/calling.html

## Out of Scope

-

## Open Questions

-

**Comment by k-macmillan** [comment]
Started drilling into this. Part of our issue was misunderstanding of immutable signatures. This is chaining without immutable:
![image.png](https://images.zenhubusercontent.com/620be841701e64d34dd34065/4e4cfbdb-86a3-4e10-aad3-ebb189ee7705)

Chaining with immutable:
![image.png](https://images.zenhubusercontent.com/620be841701e64d34dd34065/61ef1ef9-77a4-4b46-99d8-fa500c4cac33)

Immutable signatures cannot have args populated from previous values returns in the chain.

**Comment by k-macmillan** [comment]
All testing seems to point to it being related to the immutable signature. As can be seen below it utilizes the different workers to get the expected results of chaining together methods.
![image.png](https://images.zenhubusercontent.com/620be841701e64d34dd34065/14df8fe7-9a0e-4b44-9d18-f906a7286864)
If you follow the worker identifier it can be seen that both `a_list` and `b_list` flip the worker usage around, so this should be the solution. Going to increase dev to two celery tasks and give it a shot.

**Comment by mjones-oddball** [comment]
@EvanParish assigning to you per Kyle to pick up where he left off.

**Comment by EvanParish** [comment]
It looks like a fairly simple change. I'll get it tested in Dev and Staging once we have the issues there sorted out. 

I was doing my own verification of Kyle's testing above, it looks like `apply_async()` just makes it so the app isn't waiting on the current chain to finish. The chain seems to operate as expected, passing the values from the previous task to the next. If the task's signature is immutable, denoted with `.si` then that task will not accept values passed from the previous one and must have all of its own variables defined. These were observed by Kyle previously, just reiterating that I've verified these points as well.

**Comment by EvanParish** [comment]
While talking with Kyle we identified a place where the Onsite notification would not have been sent most of the time when an email is sent using the email address and made changes to correct that. 

I found out that testing the other aspects of this change is hard to do in Dev because we need to have valid inputs for MPI to recognize, or it will fail before even trying to send the Onsite notification. We can't successfully test the other ways of sending Onsite notifications unless we have valid values other than the va_profile_id. So we would need things like a valid ICN, BIRLSID, or EDIPI that can be validated through MPI in our Dev environment.

**Comment by k-macmillan** [comment]
To alleviate any confusion with the flow of onsite a [flowchart](https://github.com/department-of-veterans-affairs/vanotify-team/blob/master/Support/onsite_flowchart.md) has been added to our [Glossary](https://github.com/department-of-veterans-affairs/vanotify-team/blob/master/Support/Glossary.md).

**Comment by EvanParish** [comment]
Tom Harrison provided some possible ICN values to be used in the Dev environment. It looks like I'm getting the `422 error` I would expect to see in Dev when trying to use them on our API.

**Comment by EvanParish** [comment]
I had to fix a bug causing one of the unit tests to fail. I added the new Onsite celery task to the email path previously, but since it's no longer necessary I removed it. The code that was checking for it was still in the unit test however, causing it to fail. I removed the checks for the Onsite task and that took care of the problem.
